source/MaterialXGenGlsl/Nodes/TransformNodeGlsl.cpp

```diff
@@ -0,0 +1,112 @@
+            shader.createUniform(HwShader::PIXEL_STAGE, HwShader::PRIVATE_UNIFORMS, Type::MATRIX44, "u_worldInverseTransposeMatrix");
+        }
+        else
+        {
+            shader.createUniform(HwShader::PIXEL_STAGE, HwShader::PRIVATE_UNIFORMS, Type::MATRIX44, "u_worldMatrix");
+        }
+    }
+    else if (fromSpace == WORLD && (toSpace == MODEL || toSpace == OBJECT))
+    {
+        if (node.hasClassification(ShaderNode::Classification::TRANSFORM_NORMAL))
+        {
+            shader.createUniform(HwShader::PIXEL_STAGE, HwShader::PRIVATE_UNIFORMS, Type::MATRIX44, "u_worldTransposeMatrix");
+        }
+        else
+        {
+            shader.createUniform(HwShader::PIXEL_STAGE, HwShader::PRIVATE_UNIFORMS, Type::MATRIX44, "u_worldInverseMatrix");
+        }
+    }
+}
+
+void TransformNodeGlsl::emitFunctionCall(const ShaderNode& node, GenContext& context, ShaderGenerator& shadergen, Shader& shader_)
+{
+    HwShader& shader = static_cast<HwShader&>(shader_);
+
+    BEGIN_SHADER_STAGE(shader, HwShader::PIXEL_STAGE)
+        shader.beginLine();
+        shadergen.emitOutput(context, node.getOutput(), true, false, shader);
+
+        const ShaderInput* toSpaceInput = node.getInput(TO_SPACE);
+        string toSpace = toSpaceInput ? toSpaceInput->value->getValueString() : EMPTY_STRING;
+
+        const ShaderInput* fromSpaceInput = node.getInput(FROM_SPACE);
+        string fromSpace = fromSpaceInput ? fromSpaceInput->value->getValueString() : EMPTY_STRING;
+
+        const ShaderInput* inInput = node.getInput("in");
+
+        shader.addStr( " = ");
+        if (inInput->type == Type::VECTOR3)
+        {
+            shader.addStr("(");
+        }
+        if ((fromSpace == MODEL || fromSpace == OBJECT) && toSpace == WORLD)
+        {
+            if (node.hasClassification(ShaderNode::Classification::TRANSFORM_NORMAL))
+            {
+                shader.addStr("u_worldInverseTransposeMatrix * ");
+            }
+            else
+            {
+                shader.addStr("u_worldMatrix * ");
+            }
+        }
+        else if (fromSpace == WORLD && (toSpace == MODEL || toSpace == OBJECT))
+        {
+            if (node.hasClassification(ShaderNode::Classification::TRANSFORM_NORMAL))
+            {
+                shader.addStr("u_worldTransposeMatrix * ");
+            }
+            else
+            {
+                shader.addStr("u_worldInverseMatrix * ");
+            }
+        }
+
+        if (inInput->type == Type::VECTOR3)
+        {
+            shader.addStr("vec4(");
+            if (node.hasClassification(ShaderNode::Classification::TRANSFORM_POINT))
+            {
+                shadergen.emitInput(context, inInput, shader);
+                shader.addStr(", 1.0)");
+            }
+            else
+            {
+                shadergen.emitInput(context, inInput, shader);
+                shader.addStr(", 0.0)");
+            }
+            shader.addStr(").xyz");
+        }
+        else
```

Comment:
Might want to do an earlier sanity check for vec3 and vec4 only and fail if it's not that.
Better than failing during shader compile in case it's the wrong type.